### PR TITLE
history refactor + small feature

### PR DESCRIPTION
**Pull request by laerus**
a first try in refactoring history

sessions  are now options to the `show` subcommand i.e:
`history show bash`
`history show xonsh` for all xonsh history
`history` and `history show` implicitly means `history show session` and returns current session history as expected

history now accepts multiple slices arguments separated by spaces i.e:
`history show bash 1:3 10 11 14 15:30:5` returns the respective ranges or indexes etc.

changes in `tools.py`:
`ensure_int_or_slice` has now been deprecated for the new `ensure_slice` that always return a slice
and uses regular expressions, 

`is_int_as_str` and `is_slice_as_str` are now reimplemented in EAFP style.

### CODE REVIEW

File: tests/test_tools.py [hunk 1]
```diff
@@ -853,24 +813,37 @@ def test_bool_or_int_to_str(inp, exp):
...
+    'test',
+    '6.53:100:5',
+    '4:-',
+    '2:15-:3',
+    '50:-:666',
+    object(),
+    ])
+def test_ensure_slice_invalid(inp):
+    obs = ensure_slice(inp)
+    assert obs is None
```
Comment (scopatz) [inline tests/test_tools.py]: Failures shouldn't return None, failures should raise an error

File: xonsh/history.py [hunk 2]
```diff
@@ -357,8 +357,6 @@ def _bash_hist_parser(location=None):
                     bash_hist_formatted.append((command, 0.0, ind))
                 return bash_hist_formatted
     else:
-        import ipdb
-        ipdb.set_trace()
```
Comment (scopatz) [inline xonsh/history.py]: Thanks for catching these.

File: xonsh/history.py [hunk 3]
```diff
@@ -648,8 +617,7 @@ def show(self, *args, **kwargs):
         Valid options:
             `session` - returns xonsh history from current session
             `show`    - alias of `session`
-            `all`     - returns xonsh history from all sessions
-            `xonsh`   - alias of `all`
+            `xonsh`   - returns xonsh history from all sessions
```
Comment (scopatz) [inline xonsh/history.py]: Please retain `all`

File: xonsh/history.py [hunk 4]
```diff
@@ -680,39 +648,42 @@ def _hist_gc(ns, hist):
             continue
 
 
-_HIST_MAIN_ACTIONS = LazyObject(lambda: {
+@lazyobject
+def HIST_SESSIONS():
+    return ('session', 'xonsh', 'bash', 'zsh')
```
Comment (scopatz) [inline xonsh/history.py]: It seems that this should be a frozenset rather than a tuple

File: xonsh/history.py [hunk 5]
```diff
@@ -680,39 +648,42 @@ def _hist_gc(ns, hist):
...
     'diff': _dh_main_action,
     'gc': _hist_gc,
-    }, globals(), '_HIST_MAIN_ACTIONS')
+    }
 
 
-def _hist_main(hist, args):
-    """This implements the history CLI."""
-    if not args or args[0] in ['-r'] or ensure_int_or_slice(args[0]):
+def parse_args(args):
```
Comment (scopatz) [inline xonsh/history.py]: Due to amalgamation, please don't have such generically named funtions.  Retain the use of `_hist` in the name.  Also, all public functions need a docstring

File: xonsh/tools.py [hunk 6]
```diff
@@ -763,20 +763,6 @@ def is_int(x):
...
-    Tests if something is an integer
-    If not a string to begin with is automatically false
-    """
-    if isinstance(x, str):
-        try:
-            int(x)
-        except ValueError:
-            return False
-        return True
-    return False
```
Comment (scopatz) [inline xonsh/tools.py]: I don't see a reason to remove these functions or their tests, even if they are not used.  They could still be useful to users or us in the future for other environment variable validators.

File: xonsh/tools.py [hunk 7]
```diff
@@ -792,25 +778,6 @@ def is_slice(x):
...
-            x = x.strip('[]()')
-            try:
-                slice(*(int(x) if len(x) > 0 else None
-                        for x in x.split(':')))
-            except ValueError:
-                return False
-        else:
-            return False
-        return True
-    return False
```
Comment (scopatz) [inline xonsh/tools.py]: Same here.  Please retain the function and the tests

File: xonsh/tools.py [hunk 8]
```diff
@@ -956,19 +923,27 @@ def bool_or_int_to_str(x):
...
-    elif is_int_as_str(x):
-        return int(x)
-    return False
+        m = SLICE_REG.fullmatch(x)
+        if m:
+            groups = (int(i) if i else None for i in m.groups())
+            s = slice(*groups)
+    except TypeError:
+        pass
+    return s
```
Comment (scopatz) [inline xonsh/tools.py]: Ensurers should raise an error if they fail.

**Comment by scopatz** [comment]
Thanks for writing this @laerus!

**Comment by scopatz** [comment]
I am happy with the new functions and most of the refactor, but I think the old functions should stay too.

File: tests/test_tools.py (hunk 1 above)
Comment (laerus) [inline tests/test_tools.py]: i wanted to ask you about this, the previous `ensure_int_or_slice` returned False, i prefer raising and error too.

Comment (scopatz) [inline tests/test_tools.py]: let's do that then :)

File: xonsh/tools.py (hunk 7 above)
Comment (laerus) [inline xonsh/tools.py]: i can change this to just use the regex

Comment (scopatz) [inline xonsh/tools.py]: that would be even better

File: xonsh/tools.py (hunk 6 above)
Comment (laerus) [inline xonsh/tools.py]: this can fail with TypeError, also the docstring is misleading. *something*

**Comment by laerus** [comment]
in general these 2 functions i removed are not very well implemented imo and they existed only for the old `is_int_or_slice` which wasnt very useful either cause it required another check on the callers level. sry i rushed to a removal its just that i prefer EAFP. will bring them back though.

**Comment by laerus** [comment]
arggg flake8 complains

**Comment by scopatz** [comment]
the news test failure seems like a legitimate failure. Howeve, the flake8 indentation error seems incorrect. I think you can disable this in setup.cfg

**Comment by laerus** [comment]
fixed news and the pep8 complain but imo it seems a little more ugly.

**Comment by scopatz** [comment]
Oh yeah, I was suggesting you chabge setup.cfg to make flake8 not report those errors anymore :)

**Comment by laerus** [comment]
i think this is ready. well history could use a lot more refactoring and tests but this is enough for a start.

**Comment by scopatz** [comment]
Hi @laerus - this looks good except for the loss of the `all` type in the show command.  Please keep this around. Thanks!

**Comment by laerus** [comment]
i have included `all` too, just forgot to update the docstring of `_hist_show`, btw `all` is just an alias for `xonsh` and a little misleading, cause you'd expect to return history from *all* xonsh/bash/zsh but it just returns all xonsh history.

**Comment by scopatz** [comment]
Ahh I see.  Well I do think it needs to be in the docstrings so people know what it does.  Maybe you can change the docstring to say that all is an alias for xonsh.

**Comment by laerus** [comment]
Yes i will include it in the docs except if we want to change it so `all` really returns history from all sessions.

**Comment by scopatz** [comment]
I don't think that we need to change it in the short term

**Comment by scopatz** [comment]
That is, we don't need to change the behaviour, we just need to change the docs.

**Comment by laerus** [comment]
atm there is a small bug but i can't figure out yet where it comes from

when i type an invalid slice like `history 10:-` with `XONSH_SHOW_TRACEBACK` set to False the first time the error doesn't appear, any consecutive runs show the error msg normally though.

```
$ history 10:-
'10:-' is not a valid slice format
```

with `XONSH_SHOW_TRACEBACK` set to True this doesn't happen.

**Comment by scopatz** [comment]
Is `display_error_message()` being called? https://github.com/xonsh/xonsh/blob/master/xonsh/tools.py#L639

**Comment by laerus** [comment]
Doesn't seem so. i shouldn't call it manually right?

**Comment by scopatz** [comment]
No, you shouldn't call it manually.  This seems like an unrelated issue.

**Comment by scopatz** [comment]
Also, it seems to work just fine for me, even on the first time

**Comment by scopatz** [comment]
```console
scopatz@localhost ~/temp $ xonsh --shell-type=readline -DXONSH_COLOR_STYLE=default
scopatz@localhost ~/temp $ $XONSH_SHOW_TRACEBACK = False
scopatz@localhost ~/temp $ history 10:-
'10:-' is not a valid slice format
scopatz@localhost ~/temp $ 
```

**Comment by scopatz** [comment]
Thanks a ton @laerus!

**Pull request merged**